Ansible ssh alt

jeromew opened the pull request:
This is a PR for an alternate ssh connection_plugin following a discussion on devel ML :
- https://groups.google.com/forum/#!topic/ansible-devel/vCEZOmcLipU
- https://groups.google.com/forum/#!topic/ansible-devel/67ltJ2OGlLc

Use with -c ssh_alt

The objective is to minimize the number of ssh roundtrips for each command/module execution.
Currently when sudo'ing we often have up to 5 ssh round trips :
- make tmp dir
- change rights
- move module code
- execute module
- remove tmp dir

The new plugin brings this down to one round-trip in most cases when a tmp directory is not necessary (by piping the module source code directly over ssh to the extracted module shebang).

This implementation needs to be tested but is believed to hopefully handle most of the cases.

It should currently work in these cases :
- when -k and -K are needed
- old style & new style modules
- specific modules needing tmp dir (for example copy)
- raise an error when remote host is not reachable

Implementation details :
- during the early decision to create a tmp dir or not, a hardcoded list of module needing a tmp path: 'assemble', 'copy', 'script', 'template', 'unarchive'. This list can easily be removed by adding NEEDS_TMPPATH = True in the respective modules. I did not do it to minimize the footprint of the PR and draw attention to these early/late scenarios.
- I could not find a way to protect 3rd party module that NEED the tmp path but that do not declare it stating NEEDS_TMPPATH = True. Maybe we could add a deprecation warning and make it compulsory to declare NEEDS_TMPPATH.
- I used the already existing convention of testing tmp.find("tmp") != -1 to know whether the tmp path has been created or not (found this in many places around the codebase). Note that there is probably a bug when C.DEFAULT_REMOTE_TMP is customized without "tmp" in it. I did not try to fix this bug as I considered that it was not related.
- I observed in the codebase that some helpers in the python api related to tmp dir operations could help. some modules like copy are not very clear related to their handling of tmp path and were maybe written before persist_files ever existed.

jeromew committed:
copy ssh.py to ssh_alt.py

jeromew committed:
ssh_alt.py / decrease # of ssh roundtrips

mpdehaan commented:
Reviewing this now, thanks!

mpdehaan commented:
Ok this looks VERY good from my testing.

Before I merge this in and get everyone else to help test, I have one request about this line.
-        if module_name in ['assemble', 'copy', 'script', 'template', 'unarchive']:

Let's see if the module is an action plugin using the plugin loader.  And then, IF it is, we can do this

action_plugin.get('TRANSFERS_FILES',False)

and that way action plugins that transfer files can request the temp path that way and people won't be confused if they try to create another without knowing (because they'll copy another action plugin to start)

Does this sound reasonable?

I think once that's out of the way we can merge and get everybody to test and report comparisons.

Thanks @jeromew!  Good stuff!

jeromew commented:
I squashed the branch after making the modification to use TRANSFERS_FILES (the date of the commit was not modified)

Please note that I removed NEEDS_TMPPATH I hope that was your intent. A grep on the codebase makes me confident that if we leave this variable in it would only make additional noise for plugin developers.

I left TRANSFERS_FILES = False in files that used to have NEEDS_TMPPATH = False.

added TRANSFERS_FILES = True to ['assemble', 'copy', 'script', 'template', 'unarchive']

mpdehaan commented:
Excellent, checking this out now.

mpdehaan commented:
merged in, thanks!  I just emailed ansible-project so we can test more widely and decide whether to fold this in, which I don't see why we won't be able to.

Thanks again!

jeromew commented:
Sorry about this I forgot to test the async case.
I sent a squashed PR that handles the async case here: https://github.com/ansible/ansible/pull/5247 (github did not attach this commit here)

Just for information I will be afk for the end of the week.
